chips.asm

```diff
@@ -430,80 +430,131 @@ ALIGN SectorSize, db 0
+    dw %2               ; id
+    dd 0x50010000|%7    ; style (WS_CHILD|WS_VISIBLE|WS_TABSTOP)
+    db 0x80             ; class
+    db %1               ; text
+    db 0
+%endmacro
+
+; EDITTEXT id, x, y, width, height, [style]
+%macro EDITTEXT 5-6 0
+    dw %2, %3, %4, %5   ; position and size
+    dw %1               ; id
+    dd 0x50810000|%6    ; style (WS_CHILD|WS_VISIBLE|WS_BORDER|WS_TABSTOP)
+    db 0x81             ; class
+    db 0                ; text
+    db 0
+%endmacro
+
+; STATICTEXT text, id, x, y, width, height, [style]
+%macro STATICTEXT 6-7 0
+    dw %3, %4, %5, %6   ; position and size
+    dw %2               ; id
+    dd 0x50020000|%7    ; style (WS_CHILD|WS_VISIBLE|WS_GROUP)
+    db 0x82             ; class
+    db %1               ; text
+    db 0
+%endmacro
+
+; LISTBOX id, x, y, width, height, [style]
+%macro LISTBOX 5-6 0
+    dw %2, %3, %4, %5   ; position and size
+    dw %1               ; id
+    dd 0x50a10000|%6    ; style (WS_CHILD|WS_VISIBLE|WS_BORDER|WS_VSCROLL|WS_TABSTOP)
+    db 0x83             ; class
+    db 0                ; text
+    db 0
+%endmacro
+
+; Common styles
+%define WS_DISABLED         0x08000000
+
+; Styles for DIALOG
+%define DS_SETFONT          0x40
+%define DS_MODALFRAME       0x80
+
+; Styles for PUSHBUTTON
+%define BS_PUSHBUTTON       0
+%define BS_DEFPUSHBUTTON    1
+
+; Styles for EDITTEXT
+%define ES_UPPERCASE        0x08
+%define ES_AUTOHSCROLL      0x80
+
+; Styles for STATICTEXT
+%define SS_LEFT             0
+%define SS_CENTER           1
+%define SS_RIGHT            2
+
+; Styles for LISTBOX
+%define LBS_NOTIFY          0x01
+%define LBS_USETABSTOPS     0x80
+
 ; 3fe00
 DLGGOTO:
-    dd 0x90c800c0 ; window style
-    db 7    ; number of items
-    dw 6, 18, 151, 94 ; position and size
-    db 0x0   ; menu
-    db 0x0   ; window class
-    db "Go To Level", 0 ; caption
-    dw 8 ; font size
-    db "MS Sans Serif", 0 ; font
-    ; Items
-    db 0x09, 0x00, 0x07, 0x00, 0x84, 0x00, 0x13, 0x00, 0xff, 0xff, 0x00, 0x00, 0x02, 0x50, 0x82, "Enter a level number and password, or just a password.", 0, 0
-    db 0x45, 0x00, 0x1e, 0x00, 0x20, 0x00, 0x0c, 0x00, 0x64, 0x00, 0x80, 0x00, 0x81, 0x50, 0x81, 0, 0
-    db 0x45, 0x00, 0x2e, 0x00, 0x20, 0x00, 0x0c, 0x00, 0x65, 0x00, 0x88, 0x00, 0x81, 0x50, 0x81, 0, 0
-    db 0x10, 0x00, 0x20, 0x00, 0x33, 0x00, 0x08, 0x00, 0xff, 0xff, 0x02, 0x00, 0x02, 0x50, 0x82, "Level number:", 0, 0
-    db 0x10, 0x00, 0x2f, 0x00, 0x33, 0x00, 0x08, 0x00, 0xff, 0xff, 0x02, 0x00, 0x02, 0x50, 0x82, "Password:", 0, 0
-    db 0x1b, 0x00, 0x4a, 0x00, 0x28, 0x00, 0x0e, 0x00, ID_OK, 0x0, 0x01, 0x00, 0x01, 0x50, 0x80, "OK", 0, 0
-    db 0x53, 0x00, 0x4a, 0x00, 0x28, 0x00, 0x0e, 0x00, ID_CANCEL,0,0x00, 0x00, 0x01, 0x50, 0x80, "Cancel", 0, 0
+    DIALOG 6, 18, 151, 94, 7, {"Go To Level",0}, 8, {"MS Sans Serif",0}, DS_MODALFRAME|DS_SETFONT
+
+    STATICTEXT  {"Enter a level number and password, or just a password.",0}, -1, 9, 7, 132, 19
```

Comment:
I couldn't find STATICTEXT in the resource compiler docs. Should this be LTEXT? https://docs.microsoft.com/en-us/windows/win32/menurc/ltext-control

Also the trailing NULs in the item text should be in the macro, not here.

Reply:
> I couldn't find STATICTEXT in the resource compiler docs. Should this be LTEXT? https://docs.microsoft.com/en-us/windows/win32/menurc/ltext-control

Yes, I was being lazy and combining the three TEXT types into a single one...  I like your idea of moving the macros to a separate file though, so I can do that and be less lazy.


> Also the trailing NULs in the item text should be in the macro, not here.

I agree.  Unfortunately, I couldn't find a way to have the macro only add the trailing NUL if the string itself was non-null (several of the STATICTEXT objects have a null text entry), and I didn't really like the idea of adding a separate macro just for TEXT controls with null text.  Suggestions welcome :)